### DISCUSSION TITLE
Impossible to simulate a peg-in-hole operation using MuJoCo

### DISCUSSION THREAD

**Discussion by martcram** [google-deepmind/mujoco#738] [General]
Dear,

Yoon et al. recently concluded in their Applied Sciences journal paper titled "[Comparative Study of Physics Engines for Robot Simulation with Mechanical Interaction](https://www.mdpi.com/2076-3417/13/2/680)" that it is **impossible to simulate a peg-in-hole operation using MuJoCo**. When vertically inserting a peg in a hole with a clearance of 0.5 mm, they observed that the peg refused to move downward regardless of the amount of force applied to it. This behaviour has been attributed to the inability of MuJoCo's collision detector to handle non-convex meshes.

Without wishing to detract from the fine work of these researchers, are there other users who have made the same observation and/or found solutions to circumvent this limitation? I read about [decomposing the non-convex mesh into a union of convex meshes](https://roboti.us/forum/index.php?threads/convex-decomposition-does-not-add-concavity-to-simulation-objects.13/#post-5925) to get around this. This approach was also suggested by Yoon et al.

Thank you in advance.

Kind regards,
Martijn

**Comment by yuvaltassa** [comment]
1. That is correct, right now convex decomposition is required. This can be [automated](https://github.com/Unity-Technologies/VHACD#the-v-hacd-library-decomposes-a-3d-surface-into-a-set-of-near-convex-parts). 
2. One can also build a shape with primitives. Meshes are, even in good scenarios, rarely the most efficient representation of shape.
3. That said, there are lots of meshes in the world and we would like to be able to collide with them directly, without decomposition. We are currently in the research and prototyping stage of direct mesh collision support.
4. You may abstain from detracting, I cannot. The paper is shoddy. Example: The thing being simulated is the convex hull, they know and explain this. Yet they try to push a peg into a solid block, and make a figure showing that it doesn't penetrate? Example: the authors (and apparently their reviewers as well), did not consider that neither we nor, I can only assume, the other developers, could be contacted with questions. Example: MuJoCo's contacts are _tunable_. They can simulate steel or rubber or sponge. By default, contacts in MuJoCo have the hardness of plastic, more-or-less. The other simulators are also tunable. So the plots are basically comparing default values. A physics simulator is not like file compression software, it is like a CAD program. It has options. You need to understand how to use it, or you will misuse it.

**Comment by martcram** [comment]
Thank you for the detailed reply and for pointing out the default behaviour of MuJoCo (and the other simulators). So due to the range of tuneable parameters of each physics engine, I conclude that without configuring these correctly one is actually comparing apples and oranges. This puts the results described in the aforementioned paper in a different light. I will have a look at convex decomposition for simulating the peg-in-hole operation. This answers my question.

**Comment by hai-h-nguyen** [comment]
This video seems to be able to do it https://www.youtube.com/watch?v=PHuBPm6bNWU. @yuvaltassa  do you have any thoughts? I am having a round hole and using primitives does not work for this simple hole.

**Reply by hai-h-nguyen** [reply]
The guy in the video used the primitive approach. There is a good tool for automating the mesh decomposition approach (based on V-HACD) https://github.com/kevinzakka/obj2mjcf.

**Reply by yuvaltassa** [reply]
[This](https://github.com/SarahWeiii/CoACD) is much better. I think @kevinzakka intends to upgrade obj2mjcf at some point...

**Reply by kevinzakka** [reply]
After CoRL :)
